src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -769,10 +825,15 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
                 {
                     if (_decryptedBytesCount != 0)
                     {
+                        if (renegotiation)
+                        {
+                            throw new InvalidOperationException(SR.net_ssl_renegotiate_data);
```

Comment:
What state is the SslStream left in after this occurs?

From looking at the code, it seems like the following will happen currently:
(1) NegotiateClientCertificateAsync will throw
(2) Subsequent Read calls on this SslStream will succeed with whatever data we had or receive in the future
(3) If we ever do receive a response to the renegotiation, we will process it and make the cert available, if any

(3) in particular seems weird.

Reply:
yes, the description is right. The (3) can conceptually already happen if renegotiation is triggered by other side and peer sends certificates for what ever reason. We can probably detect such state and simple ignore the provided late certificate. Let me know if I should code that. It would probably be around VerifyRemoteCerttificate code and it would probably work same way for Linux.

Reply:
It seems like we shouldn't ever process a client certificate if we aren't expecting it to be delivered by the client... i.e. initial handshake or negotiation caused by us calling NegotiateClientCertificateAsync. Otherwise you could get into some very weird behavior, like a certificate showing up randomly. What do you think?

Reply:
Also... if we let you continue to read/write from the stream after NegotiateClientCertificateAsync fails, then isn't there a possibility of getting into a deadlock later?

Reply:
I added code to ignore remote certificates received outside of handshake and renegotiation requested by us.

I don't think there is deadlock in ssl stream. The one @Tratcher mentioned is caused by trying to renegotiate while not reading. All the stacks are already capable of consuming renegotiation / handshake frames during read. So as long as somebody is trying to read from the stream, rest should work just fine.